File: apps/api/src/lib/s3.service.ts
import { Injectable, Optional } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import { AppConfiguration } from '@/config/configuration';

@Injectable()
export class S3Service {
  private readonly client: S3Client;
  private readonly bucketName: string;

  constructor(
    @Optional()
    private readonly configService?: ConfigService<AppConfiguration>,
  ) {
    const endpoint =
      this.configService?.get('r2Endpoint', { infer: true }) ||
      'https://localhost';
    const accessKeyId =
      this.configService?.get('r2AccessKeyId', { infer: true }) || 'dummy';
    const secretAccessKey =
      this.configService?.get('r2SecretAccessKey', { infer: true }) || 'dummy';
    this.bucketName =
      this.configService?.get('r2BucketName', { infer: true }) || 'dummy';

    this.client = new S3Client({
      region: 'auto',
      endpoint,
      forcePathStyle: true,
      credentials: {
        accessKeyId,
        secretAccessKey,
      },
    });
  }

  async uploadFile(
    fileBuffer: Buffer,
    key: string,
    contentType: string,
  ): Promise<string> {
    return this.uploadImage(key, fileBuffer, contentType);
  }

  async uploadImage(
    key: string,
    fileBuffer: Buffer,
    contentType: string,
  ): Promise<string> {
    const command = new PutObjectCommand({
      Bucket: this.bucketName,
      Key: key,
      Body: fileBuffer,
      ContentType: contentType,
    });

    await this.client.send(command);

    return key;
  }
}

File: apps/api/services/s3_service.go
package services

import (
	"bytes"
	"fmt"
	"net/http"
	"strings"
	"time"

	"gaeldle/api/config"
	"gaeldle/api/lib"
)

type S3Service struct {
	endpoint        string
	bucketName      string
	accessKeyID     string
	secretAccessKey string
	region          string
	client          *http.Client
}

func NewS3Service(cfg *config.AppConfig) *S3Service {
	endpoint := cfg.R2Endpoint
	if endpoint == "" {
		endpoint = "https://localhost"
	}
	region := cfg.AwsRegion
	if region == "" {
		region = "auto"
	}
	return &S3Service{
		endpoint:        strings.TrimRight(endpoint, "/"),
		bucketName:      cfg.R2BucketName,
		accessKeyID:     cfg.R2AccessKeyID,
		secretAccessKey: cfg.R2SecretAccessKey,
		region:          region,
		client:          &http.Client{Timeout: 30 * time.Second},
	}
}

func (s *S3Service) UploadImage(key string, fileBuffer []byte, contentType string) (string, error) {
	if s.bucketName == "" || s.accessKeyID == "" || s.secretAccessKey == "" {
		return "", fmt.Errorf("R2/S3 credentials or bucket name not configured")
	}

	trimmedKey := strings.TrimLeft(key, "/")
	targetURL := fmt.Sprintf("%s/%s/%s", s.endpoint, s.bucketName, trimmedKey)

	req, err := http.NewRequest("PUT", targetURL, bytes.NewReader(fileBuffer))
	if err != nil {
		return "", fmt.Errorf("failed to create upload request: %w", err)
	}

	if contentType == "" {
		contentType = "image/jpeg"
	}
	req.Header.Set("Content-Type", contentType)

	now := time.Now().UTC()
	if err := lib.SignRequest(req, fileBuffer, "s3", s.region, s.accessKeyID, s.secretAccessKey, now); err != nil {
		return "", fmt.Errorf("failed to sign upload request: %w", err)
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to execute upload request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusNoContent {
		return "", fmt.Errorf("R2/S3 upload returned status %d", resp.StatusCode)
	}

	return key, nil
}
